### ISSUE TITLE
double precision produces different results in lammps 2022 and lammps 2021

### ISSUE THREAD

**Issue by Lewis-YL** [lammps/lammps#3580]
**Summary**

Double precision calculations in LAMMPS 2022 and LAMMPS 2021 occasionally lose digits in energy and these errors accumulate and eventually result in different output.

Is there a way to avoid this digits lost during calculation?

**LAMMPS Version and Platform**

LAMMPS (23 Jun 2022) on Ubuntu 22.04
LAMMPS (29 Sep 2021 - Update 3) on Ubuntu 22.04

**Details**

LAMMPS (29 Sep 2021 - Update 3)
![LAMMPS 2021](https://user-images.githubusercontent.com/38740371/209885759-2aee652a-4d8f-4a07-8556-66bc52599c00.png)

LAMMPS (23 Jun 2022) 
![LAMMPS 2022](https://user-images.githubusercontent.com/38740371/209885780-33c8b964-9182-4ef7-b53e-cde086c96b5a.png)

**Comment by akohlmey** [comment]
> Is there a way to avoid this digits lost during calculation?

Not really. MD is a chaotic system, so even the tiniest changes will eventually grow (exponentially) so that trajectories will diverge. The LAMPS documentation mentions that: https://docs.lammps.org/Errors_common.html

At the core this is due to using floating point math and the fact that it is not associative, i.e. the exact result depends on the order of operations.  Some detailed discussions of this is on these websites, for example: 
http://blog.reverberate.org/2014/09/what-every-computer-programmer-should.html
http://blog.reverberate.org/2016/02/06/floating-point-demystified-part2.html
http://floating-point-gui.de/

The only way to avoid this is to write an MD code that uses fixed point math (typically via scaled integers).

Items that can trigger this divergence are: a different number of MPI processes, different neighbor list settings, different compiler, different compiler version, different versions of external libraries (e.g. libm), different compiler settings/optimization, changes in the source code to add features or remove bugs (may influence how the compiler can optimize). How fast this divergence occurs depends on a number of factors, too. Particularly high impact have pair styles that use spline functions or any tabulation,  triclinic simulation cells, and using fixes that change the volume of the simulation cell.

Bottom line: you can only expect identical results with identical executables on the exact same hardware and operating system.

**Comment by Lewis-YL** [comment]
This is very informative.

Thank you so much!
